app/src/main/java/org/gdg/frisbee/android/gde/GdeActivity.java

```diff
@@ -94,102 +84,66 @@ public void networkFailure(Throwable error) {
+        HashMap<String, GdeList> gdeMap = extractCategoriesFromGdeList(directory);
+        List<GdeCategory> gdeCategoryList = convertCategoryMapToList(gdeMap);
+
+        mViewPagerAdapter = new GdeCategoryPagerAdapter(
+            getSupportFragmentManager(),
+            getString(R.string.about),
+            gdeCategoryList
+        );
+        mViewPager.setAdapter(mViewPagerAdapter);
+        mViewPager.addOnPageChangeListener(this);
+        mTabLayout.setupWithViewPager(mViewPager);
+
+        getAchievementActionHandler().handleLookingForExperts();
     }
 
-    public class GdeCategoryAdapter extends FragmentStatePagerAdapter implements ViewPager.OnPageChangeListener {
-        private final SparseArray<WeakReference<Fragment>> mFragments = new SparseArray<>();
-        private HashMap<String, GdeList> mGdeMap;
-
-        public GdeCategoryAdapter(FragmentManager fm) {
-            super(fm);
-            mGdeMap = new HashMap<>();
-        }
-
-        public void addMap(Map<String, GdeList> collection) {
-            mGdeMap.putAll(collection);
-        }
-
-        @Override
-        public int getCount() {
-            return mGdeMap.keySet().size() + 1;
-        }
-
-        @Override
-        public Fragment getItem(int position) {
-            if (position == 0) {
-                return PlainLayoutFragment.newInstance(R.layout.fragment_gde_about);
-            } else {
-                String key = mGdeMap.keySet().toArray(new String[mGdeMap.size()])[position - 1];
-                Fragment frag = GdeListFragment.newInstance(mGdeMap.get(key));
-                mFragments.append(position, new WeakReference<>(frag));
+    private HashMap<String, GdeList> extractCategoriesFromGdeList(GdeList directory) {
+        HashMap<String, GdeList> gdeMap = new HashMap<>();
 
-                return frag;
+        for (Gde gde : directory) {
+            if (gde.getProduct() == null) {
+                continue;
             }
-        }
+            for (String p : gde.getProduct()) {
+                String product = p.trim();
+                if (!gdeMap.containsKey(product)) {
+                    gdeMap.put(product, new GdeList());
+                }
 
-        @Override
-        public CharSequence getPageTitle(int position) {
-            if (position == 0) {
-                return getString(R.string.about);
-            } else if (position > -1 && position - 1 < mGdeMap.keySet().size()) {
-                String title = mGdeMap.keySet().toArray(new String[mGdeMap.size()])[position - 1];
-                title = title.length() > 14 ? Utils.getUppercaseLetters(title) : title;
-                return title;
-            } else {
-                return "";
+                gdeMap.get(product).add(gde);
             }
         }
+        return gdeMap;
+    }
 
-        @Override
-        public void onPageScrolled(int i, float v, int i2) {
+    private List<GdeCategory> convertCategoryMapToList(HashMap<String, GdeList> gdeMap) {
```

Comment:
Using `HashMap` in the adapter is also really really bad. I was wondering about the ordering of the categories in the ViewPager titles. The ordering was random. I think that is also because of this HashMap. 

I can move this convertion code to a background thread.